features/device-config/monitor-adaptor/src/main/java/org/opennms/features/deviceconfig/monitor/adaptor/DeviceConfigMonitorAdaptor.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.deviceconfig.monitor.adaptor;
+
+import com.google.common.base.Strings;
+import org.opennms.features.deviceconfig.persistence.api.ConfigType;
+import org.opennms.features.deviceconfig.persistence.api.DeviceConfig;
+import org.opennms.features.deviceconfig.persistence.api.DeviceConfigDao;
+import org.opennms.netmgt.dao.api.IpInterfaceDao;
+import org.opennms.netmgt.model.OnmsIpInterface;
+import org.opennms.netmgt.poller.MonitoredService;
+import org.opennms.netmgt.poller.PollStatus;
+import org.opennms.netmgt.poller.ServiceMonitorAdaptor;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Map;
+import java.util.Optional;
+
+public class DeviceConfigMonitorAdaptor implements ServiceMonitorAdaptor {
+
+    private static final String DEVICE_CONFIG_MONITOR = "DeviceConfig";
+
+    @Autowired
+    private DeviceConfigDao deviceConfigDao;
+
+    @Autowired
+    private IpInterfaceDao ipInterfaceDao;
+
+    @Override
+    public PollStatus handlePollResult(MonitoredService svc, Map<String, Object> parameters, PollStatus status) {
+
+        if (!svc.getSvcName().equals(DEVICE_CONFIG_MONITOR)) {
+            return status;
+        }
+        // Retrieve interface
+        final OnmsIpInterface ipInterface = ipInterfaceDao.findByNodeIdAndIpAddress(svc.getNodeId(), svc.getIpAddr());
+        String encodingAttribute = getObjectAsString(parameters.get("encoding"));
+        String configTypeAttribute = getObjectAsString(parameters.get("config-type"));
+        String encoding = !Strings.isNullOrEmpty(encodingAttribute) ? encodingAttribute : Charset.defaultCharset().name();
+        ConfigType configType = !Strings.isNullOrEmpty(configTypeAttribute) ? ConfigType.valueOf(configTypeAttribute) : ConfigType.Default;
+
+        byte[] deviceConfigBytes = status.getDeviceConfig();
+        // Handle config retrieval failure.
+        if (deviceConfigBytes == null) {
+            Date currentTime = new Date();
+            Optional<DeviceConfig> configOptional = deviceConfigDao.getLatestConfigForInterface(ipInterface, configType);
+            DeviceConfig deviceConfig;
+            if (configOptional.isPresent() && configOptional.get().getLastFailed() != null) {
```

Comment:
Why do you check for `lastFailed != null`?

```suggestion
            if (configOptional.isPresent()) {
```